modules/kvformat/linux-audit-scanner.c

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2015 BalaBit
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+#include "kv-scanner.h"
+#include "misc.h"
+#include "utf8utils.h"
+
+#include <string.h>
+#include <ctype.h>
+
+const gchar *hexcoded_fields[] =
+{
+  "name",
+  "proctitle",
+  "path",
+  "dir",
+  "comm",
+  "ocomm",
+  "data",
+  "old",
+  "new",
+  NULL
+};
+
+static gint
+_decode_xdigit(gchar xdigit)
+{
+  if (xdigit >= '0' && xdigit <= '9')
+    return xdigit - '0';
+  xdigit = toupper(xdigit);
+  if (xdigit >= 'A' && xdigit <= 'F')
+    return xdigit - 'A' + 10;
+  return -1;
+}
+
+static gint
+_decode_xbyte(gchar xdigit1, gchar xdigit2)
+{
+  gint nibble_hi, nibble_lo;
+
+  nibble_hi = _decode_xdigit(xdigit1);
+  nibble_lo = _decode_xdigit(xdigit2);
+  if (nibble_hi < 0 || nibble_lo < 0)
+    return -1;
+  return (nibble_hi << 4) + nibble_lo;
+}
+
+static gboolean
+_is_control_char(gint ch)
+{
+  if (ch < 0x21 || ch > 0x7e)
+    return TRUE;
+  return FALSE;
+}
+
+static gboolean
+_parse_linux_audit_hexstring(GString *decoded_value, const gchar *value, gsize len)
+{
+  gint src;
+  gboolean kernel_would_have_encoded_this_as_hex = FALSE;
+
+  for (src = 0; src < len; src += 2)
+    {
+      gint v;
+
+      v = _decode_xbyte(value[src], value[src + 1]);
+
+      if (v < 0)
+        return FALSE;
+
+      if (_is_control_char(v) || v == '"')
+        kernel_would_have_encoded_this_as_hex = TRUE;
+
+      if (v == 0)
+        v = 9;
```

Comment:
because the proctitle element contains zero bytes, which we don't really
handle. I didn't want to replace that with spaces as it can render it
completely unusable, tab is a bit better as arguments - usually - don't
contain tabs. I was looking for a unicode non-breakable-space character,
but unfortunately that's not rendered properly in terminals (and neither in
a browser), so I sticked to tabs instead.

I have added a patch to convert 9 to '\t', that's a bit more readable.

-- 
Bazsi

On Mon, Sep 28, 2015 at 3:00 PM, Tibor Benke <notifications@github.com>
wrote:

> In modules/kvformat/linux-audit-scanner.c
> <https://github.com/balabit/syslog-ng/pull/715#discussion_r40548330>:
>
> > +  gboolean kernel_would_have_encoded_this_as_hex = FALSE;
> > +
> > +  for (src = 0; src < len; src += 2)
> > +    {
> > +      gint v;
> > +
> > +      v = _decode_xbyte(value[src], value[src + 1]);
> > +
> > +      if (v < 0)
> > +        return FALSE;
> > +
> > +      if (_is_control_char(v) || v == '"')
> > +        kernel_would_have_encoded_this_as_hex = TRUE;
> > +
> > +      if (v == 0)
> > +        v = 9;
>
> Why do we convert 0s to tab characters? Shouldn't we use constants here
> too?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/715/files#r40548330>.
>